### ISSUE TITLE
`active_record.instantiation` does not take `service_name` into account

### ISSUE THREAD

**Issue by beauraF** [DataDog/dd-trace-rb#2885]
**Current behaviour**
`active_record.instantiation` operation does not take `service_name` into account .

**Expected behaviour**
`active_record.instantiation` operation should respect `service_name`.

**Steps to reproduce**

```ruby
Datadog.configure do |config|
  config.service = 'default'
  config.tracing.instrument(:rails, service_name: 'rails')
  config.tracing.instrument(:active_record, service_name: "database")
end
```

With this configuration `active_record.instantiation` service name will be `default` when we should expect `database`.

**Environment**

* **ddtrace version:** 1.11.1
* **Ruby version:** 3.2.2

**Comment by TonyCTHsu** [comment]
👋 @beauraF, may I ask what would you find useful to assign every integration as a service? 

Technically speaking, `rack`, `controller` and `active_record.instantiation` all should belong to a single process/service, which is your rails app.  Our active record instrumentation would provide another `active_record.sql` span described the span for database, which would have another service name assigned.
  
We are currently planning to migrate away from this confusing `service` naming settings for each integration. I would advise to remove those `service_name` in your configuration.

```
Datadog.configure do |config|
  config.service = 'default'
  config.tracing.instrument(:rails)
  config.tracing.instrument(:active_record)
end
```

Does that make sense?

**Comment by paniko0** [comment]
@TonyCTHsu we are having a similar issue, but with a different effect. 

If you take the following very standard Rails application configuration, you'd imagine that you'd see `active_record.instantiation` data under each of the individual configurations but you won't.

```
Datadog.configure do |c|
    c.service = 'default'

    c.tracing.instrument :rails, service_name: 'web'
    c.tracing.instrument :sidekiq, service_name: 'sidekiq'
end
```

It makes us get Sidekiq instrumentation data in the default service, which is very confusing for us.

![image](https://github.com/DataDog/dd-trace-rb/assets/928351/f867b202-a6a2-4c4b-b286-b46085a47f39)

So the plan was to move this instantiation data to its own `service_name`, just like OP is trying to do, so we can get this instantiation data in a single place.

```
Datadog.configure do |c|
    c.service = 'default'

    c.tracing.instrument :rails, service_name: 'web'
    c.tracing.instrument :sidekiq, service_name: 'sidekiq'

    config.tracing.instrument(:active_record, service_name: 'ar-instrumentation')
end
```

But we reached the same point OP has where the service name is not supported on the active_record instrumentation.

> We are currently planning to migrate away from this confusing service naming settings for each integration. I would advise to remove those service_name in your configuration.

What's going to be the alternative then? Will all services live under the same name?

**Comment by delner** [comment]
Hey @paniko0, going to jump in here. Can you help me get a clearer picture of what the desired outcome is in the UI?

Are we trying to isolate the `active_record.instantiation` spans from the other instrumentation? If so for what reason? Merely to keep them segregated? Or to provide some useful grouping?

We're actually in the process of trying to provide some alternatives; if I can get a better idea of what we're aiming for here, then we may be able to provide something :)

**Comment by delner** [comment]
Separately, I'd like to provide some more context into what @TonyCTHsu is alluding to:

--- 

For some time now, the `service` field has been the only way to group spans with unique colors/stats from other spans in the UI. Because of this, its value has been often manipulated to create unique groups of spans within the application.

The big downside of this practice is `service` is meant to refer to the "name of the application that produced the span." If the same application produces spans with many different services, it fragments that application into many tiny services that makes for a bad experience (hard to navigate, reason about.)

Our goal is to undo this practice of fragmenting `service`, by having it only describe the application name (e.g. `DD_SERVICE`), thereby creating more consistency in data and experience. However, we know doing so would remove the ability "group" spans within an application (e.g. by controller, or renderer), which is crucial to many users' workflows. So we're developing an alternative, equivalent means to by which to "group" and recreate the same colorization/stat breakdowns.

In broad strokes, the current idea is pretty straightforward:

 - All spans generated in the application will always reflect `service` as the application's name: typically `DD_SERVICE` or `c.service` (whichever is configured.)
 - All spans that connect directly to an external service (e.g. HTTP request, DB query) will always have a `peer.service` tag which will reflect the name of the external service: typically this will be the name you configure per integration (e.g. `c.tracing.instrument :active_record, service_name: 'my-db'`)
 - All other spans (that do not directly connect to an external service) will only reflect the `service` field, and will never be attributed to another service. Instead, we will add a feature in the UI to group on a new tag; one that represents a useful "group name" (possibly package name, e.g. `active_record`.) The exact behavior is still TBD, and may change.
 
 ---

So what does this mean for the `active_record.instantiation`? Basically:

 1. Its service name will always be the application name (because it is a behavior internal to Ruby.) Therefore when viewing a trace by service name, it will show as a part of the same application (same color as the trace's local root span.)
 2. Its "group" tag would presumably reflect `active_record`. In dropdown within the trace graph, you would be able to group by this "group" tag, to recolorize the graph and show this span in its own group (with other ActiveRecord spans) separate from the other parts of your application.

Does this make sense? Would this be satisfactory for your use case? If not, what might I be missing?

**Comment by paniko0** [comment]
@delner Hello, and thanks for getting back to us.

> Can you help me get a clearer picture of what the desired outcome is in the UI?

TLDR; The image I posted a few weeks back exemplifies that we see instantiation data initiated by either `sidekiq` or `rails` service names in the `default` service name. We wanted to move these instantiation records to their own `instantiation` service name so they don't get mixed with the other instrumentation data that are not grouped by `sidekiq` or `rails`.

-----

The instantiation instrumentation is something that, in my opinion, belongs to whenever the place that active_record objects have been instantiated–that is, it could have been initiated as part of a web request through Rails, a worker through Sidekiq, a rake task, etc.

When we set the `service_name` as part of the Rails instrumentation (eg.: `c.tracing.instrument :rails, service_name: 'web'`), what I expect to see is active_record.instantiation values that got triggered by the Rails instrumentation to show up in the service called `web` since that is the place that the instrument is originated from. The same would, in my understanding, apply when we set the service name as part of the Sidekiq instrumentation (eg.: `c.tracing.instrument :sidekiq, service_name: 'workers'`) where I'd see active_record.instantiation values initiated by workers to show up in the service called `workers`.

However, our understanding isn't correct. We noticed that we get instantiation values on the default service name regardless of where these instantiations have occurred. This is somewhat confusing to us since we try to get spans organized by the framework that initiated them–with the exception of Postgres and Elasticsearch queries that, even though are initiated by Rails or Sidekiq, they are already going to the default service name, which basically forced us to think of it as, in very simple terms, "things that run outside of the context of the app".

To summarize, I think these are our expectations and misconceptions on how to use Datadog:
- We expect to see data based on the service that initiated them
  1. This is incorrect since the traces are organized by service, and not by each individual instrumentation object
- We want to see the lifecycle of a request
  1. We don't _really_ care about individual instantiation information. They are **very important** for us when we open a slow request and look at the flame graph, but they give us very little value when they are individually displayed on the table list in the default service name. 
- Perhaps the main misconception on my part is that I expect the Datadog's Traces to work like many other APM tools that I use in the past: I can filter between web requests and workers, then check out of everything that is running, what is initiating too many AR objects, or which requests are taking the longest, or which workers are running the most, etc.
  1. We kind of simulate this behaviour by making use of individual `service_name` per instrumentation, but based on your message and many other that I was part of with Datadog's support, this is something that is in the process of being deprecated. 

Again, thanks for getting back to us and I hope we have give enough clarity on what/how we were expecting this to work.

**Comment by delner** [comment]
@paniko0 Thank you for the thorough explanation! This really helps clarify expectations a lot for me.

I think the expected behavior you're laying out makes a lot of sense: keep behavior attributed to the process that performed it, not to some individual category. I think we could make it work this way.

Still, I need to clarify some existing behavior vs desired behavior:

#### Existing behavior

Right now, it's not uncommon for one web request in one Rails app to generate spans for multiple services (e.g. a request that queues a job creates spans for `web` --> `workers`.)

  - This is to say, one Ruby process is modeled as multiple "services"

#### Issue & rationale

Having multiple services per app fractures one app into many. We don't believe one Ruby process should create spans that belong to multiple services. Service name should be analogous to application name.

#### Desired behavior

All spans generated from one Ruby process should all bear the same service on every span.

### Concerning instantiation...

Our current service behavior is a problem for your desired instantiation behavior because what service this instantiation span should be attributed to. Spans do not see the parent's service. Moreover, even if it can, the parent service is a naive choice, depending on when the instantiation occurs in the trace. So we always choose the default service.

I think your desired instantiation behavior is achievable not by making instantiation derive some other service name, but by having all spans use a single service name... this kills two birds with one stone by (1) consolidating operations under one service that represents the app [which is easier to understand/work with], and (2) negating the need to do any kind of logic to reassign services.

The only downside (that I see) with this is that "service" is sometimes used to create distinct layers (e.g. your `web` and `worker`) within the same process, and that coloration/table entry/statistics for this category is useful. We hope to recreate this UI effect not through use of service, but through new tags.

We're starting by tagging externally facing spans (HTTP requests, DB queries etc) with `peer.service` which will allow you reproduce the original look of the flamegraph in the UI.

#### In summation...

For your example of `web` and `workers`...

- In the `web` process (e.g. `rails s`)
  - You would configure:
    ```ruby
    ENV['DD_SERVICE'] = 'web'

    Datadog.configure do |c|
      c.tracing.instrument :rails
      c.tracing.instrument :sidekiq
      c.tracing.instrument :redis, service_name: 'sidekiq_queue'
    end
    ```
  - All spans, including the instantiation span, would be mapped to `web`
  - A `sidekiq.push` span that queued a Sidekiq job would have `service = web`
  - A `redis.command` span that queued a Sidekiq job would have `service = web, peer.service = 'sidekiq_queue'`
- In the `workers` process (e.g. `bundle exec sidekiq`)
  - You would configure:
    ```ruby
    ENV['DD_SERVICE'] = 'workers'

    Datadog.configure do |c|
      c.tracing.instrument :rails
      c.tracing.instrument :sidekiq
      c.tracing.instrument :redis, service_name: 'sidekiq_queue'
    end
    ```
  - All spans, including the instantiation span, would be mapped to `workers`
  - A `sidekiq.job_fetch` span that fetched a Sidekiq job would have `service = workers`
  - A `redis.command` span that fetched a Sidekiq job would have `service = workers, peer.service = 'sidekiq_queue'`
  - A `sidekiq.job` span that performed a Sidekiq job would have `service = workers`

Then in the UI, the default view (color by service) would show all spans from the `web` process (including instantiation) in one color, all spans from `workers` (including instantiation) in another color.

If you changed the view to display by peer service, you would see just the `redis.command` spans recolored into a 3rd color and grouping instead (based on their `peer.service = 'sidekiq_queue'` tag.)

All in all, does that seem desirable?

**Comment by delner** [comment]
Hey @beauraF! I haven't seen any activity on this for a little while. Given the current behavior is intentional, and we outlined plans for future improvements above, I don't believe there is a bug here, so I'm going to close this in the interim.

If you'd like to explore the use case more with us, we invite you to continue to comment here! Thank you!
